QAOA Example with MaxCut

Gopal-Dahale opened the pull request:
## Purpose ✨

QAOA Example with MaxCut Jupyter notebook 
## Details 📝

- Describe the max cut quadratic program, and the creation of Hamiltonian
- Show an example with a smaller graph.
    - creation of the QAOA circuit
    - taking care of parameters (you remember every gate in Pauli prop was considering an independent parameter)
- optimization
- then scale it to the IBM hex topology.

## Dependencies 🔗

Graphs, GraphRecipes, Plots, JuMP, SCIP, Optim, ForwardDiff

## Future Improvements 🔭

- Possible use of symmetry merging.
- `overlapwithplus` vs `overlapwithzero`.
- cache parameter expansion array

## Mentions 👀

@MSRudolph

Gopal-Dahale committed:
QAOA example

MSRudolph commented:
I had taken a look and there are several things I would like to try for QAOA, I just haven't really gotten the time to try them. 

The first one is trying deeper circuits and different truncations, and the other is using automatic differentiation libraries. For QAOA in particular, we could use forward diff very well, and compare that with compiled gradient variants.

MSRudolph committed:
Merge pull request #122 from MSRudolph/dev

v0.5.2

MSRudolph committed:
Merge pull request #126 from MSRudolph/dev

v0.6.0

MSRudolph committed:
Merge pull request #127 from MSRudolph/dev

fix symmetry notebooks

MSRudolph committed:
Merge pull request #129 from MSRudolph/dev

Docs for symmetry merging (#128)
Add reference to new symmetry merging paper to README and example notebook

MSRudolph committed:
v0.7.0

## New Features
- `VectorPauliSum` propagation which is natively multi-threaded in propagation.
- An experimental `PauliPropagationCUDA` extension that allows PauliRotation gates to execute on the GPU.
- Schrödinger picture propagation for some gates, including `PauliRotation`, `CliffordGate`, and `PauliNoise`.
- ImaginaryPauliRotation gates in the Schrödinger picture.


## Design changes
- Large chunks of the propagation code was migrated into PauliPropagation.PropagationBase in preparation for moving it into its own PropagationBase.jl repo.
- Introduced `AbstractTermSum` and `AbstractPauliSum` as abstract types with uniting behavior. Useful defaults available for Dict and Vector storage.
- Introduced propagation caches and the abstract type `AbstractPropagationCache`. These wrap evolving `AbstractTermSum` objects with optional additional objects necessary for efficient propagation.


## Breaking Changes 
Breaking changes do not affect high-level users. Only users defining custom gates and propagation subroutines.
- The argument order and argument types of `applymergetruncate!()` and `applytoall!()` was changed for to receive propagation caches and the optional parameter as last argument.
- `applyandadd!()` was removed.
- Merging functions `mergeandempty!()` and below and truncation functions `checktruncationsonall!()` and below where simplified and reworked into `merge!()` and `truncate!()`.

MSRudolph committed:
Merge pull request #136 from MSRudolph/dev

Performance fixes:
- dispatch to Base.iterate() for PauliSum
- replace Pauli rotation gate with gate bitmask in commutes()
- inline iterate(), add!(), sett!()

Gopal-Dahale committed:
Merge branch 'main' into qaoa_example

Gopal-Dahale committed:
updated with forward diff
